### PR TITLE
Added support for depthwise separable convolutions

### PR DESCRIPTION
**Pull request by tehtea**
Hello, over the past year I've been working on-and-off using this repository as a base for my own superresolution project. During this time I've made a few changes to it which I hope will be helpful for others. Here is the summarized list of changes:
- added depthwise separable convolutional layers (can be passed as a model argument for building the graph)
- added bottleneck layers (can be passed as a model argument as well)
- changed tfGraph to inherit from tf.graph
- replaced the library to use for saving images from scipy to imageio (to solve forward compatibility issues)
- added a script for optimizing the model for inference (copied from elsewhere but placed here for convenience sake)
- added a script to convert the model to tensorflow-lite
- evaluate.py now prints the timing for performing inference per image.

Let me know what you think! Let me know if anything should be edited as well :)

### CODE REVIEW

**Comment by jiny2001** [comment]
Hi tehtea, thank you so much for the request. This looks nice!!
Please wait for somedays to be merged.  I'm very interested and really want to review each changes but the diff is large. Thx!!

File: augmentation.py [hunk 1]
```diff
@@ -9,6 +9,8 @@
 
 --augment_level 2-8: will generate flipped / rotated images
 
+Don't move this script or it can't import from helper.
```
Comment (jiny2001) [inline augmentation.py:12]: Do we need this comment? I think it's very obvious that you shouldn't move the file which is referenced from others.

File: evaluate.py [hunk 2]
```diff
@@ -64,6 +65,8 @@ def main(not_parsed_args):
         for test_data in test_list:
             evaluate_model(model, test_data)
 
+    model.save_model(name="inference_model", output_log=True, eval=False)
```
Comment (jiny2001) [inline evaluate.py]: evaluate.py is just for evaluating the model performance. So the model itself won't be changed. Should we save the model here?

File: evaluate.py [hunk 3]
```diff
@@ -80,19 +83,17 @@ def evaluate_bicubic(model, test_data):
...
-    total_psnr = total_ssim = 0
+    total_psnr = total_ssim = total_time = 0
 
     for filename in test_filenames:
-        if FLAGS.save_results:
-            psnr, ssim = model.do_for_evaluate_with_output(filename, output_directory=FLAGS.output_dir,
-                                                           print_console=False)
-        else:
-            psnr, ssim = model.do_for_evaluate(filename, print_console=False)
+        psnr, ssim, elapsed_time = model.do_for_evaluate(filename, output_directory=FLAGS.output_dir,
```
Comment (jiny2001) [inline evaluate.py:89]: Can we keep FLAGS.save_results option here? Sometimes we would like to write all the test results to the file.

File: frozenInference.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+"""
+Paper: "Fast and Accurate Image Super Resolution by Deep CNN with Skip Connection and Network in Network"
```
Comment (jiny2001) [inline frozenInference.py:2]: Can we use freeze_inference_graph.py or something like that for the filename?

Comment (jiny2001) [reply frozenInference.py:2]: Also this file is so sillier with evaluate.py. If the purpose of this file is for creating frozen graph, let's create simple python script.

File: helper/utilty.py [hunk 5]
```diff
@@ -15,7 +15,7 @@
 import tensorflow as tf
 from PIL import Image
 from os.path import isfile, join
-from scipy import misc
+import imageio as misc
```
Comment (jiny2001) [inline helper/utilty.py:18]: Please use scipy package for image writing. As far as I tested, image.imwrite() automatically rescale the image intensity and it brakes evaluation/inference. We need to use spicy without rescaling.

File: helper/utilty.py [hunk 6]
```diff
@@ -383,9 +395,10 @@ def upscale_weight(scale, channels, name="weight"):
 
 
 def weight(shape, stddev=0.01, name="weight", uniform=False, initializer="stddev"):
+    # initialize the weights of the graph as a tf variable ( to ensure non-symmetry )
     if initializer == "xavier":
         initial = xavier_cnn_initializer(shape, uniform=uniform)
-    elif initializer == "he":
+    elif initializer == "he": #this is actually the default one because args.py default argument for initializer is "he".
```
Comment (jiny2001) [inline helper/utilty.py:401]: This function is a utility function. I think these kind of comments should go to model generating function.

File: helper/utilty.py [hunk 7]
```diff
@@ -517,10 +530,19 @@ def compute_psnr_and_ssim(image1, image2, border_size=0):
...
+    
 
-    psnr = compare_psnr(image1, image2, data_range=255)
-    ssim = compare_ssim(image1, image2, win_size=11, gaussian_weights=True, multichannel=True, K1=0.01, K2=0.03,
-                        sigma=1.5, data_range=255)
+    try:
+        psnr = compare_psnr(image1, image2, data_range=255)
+        ssim = compare_ssim(image1, image2, win_size=11, gaussian_weights=True, multichannel=True, K1=0.01, K2=0.03,
+                            sigma=1.5, data_range=255)
+    except ValueError:
```
Comment (jiny2001) [inline helper/utilty.py:539]: For this case, I guess we should check numpy array type before  compare_psnr() and branch it if the type is not suitable?

File: tools/getNodeNames.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import tensorflow as tf
```
Comment (jiny2001) [inline tools/getNodeNames.py:1]: Can we use get_nod_names.py for the filename?

File: .vscode/settings.json [hunk 9]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (jiny2001) [inline .vscode/settings.json:1]: Please remove this personal settings for your vscode.

File: DCSCN.py [hunk 10]
```diff
@@ -16,6 +16,7 @@
 
 import numpy as np
 import tensorflow as tf
+from tensorflow.python.client import timeline
```
Comment (jiny2001) [inline DCSCN.py:19]: What is this package used for?

File: models/checkpoint [hunk 11]
```diff
@@ -1,2 +1,2 @@
-model_checkpoint_path: "dcscn_L2_F4to4_PS_R1F4.ckpt"
-all_model_checkpoint_paths: "dcscn_L2_F4to4_PS_R1F4.ckpt"
+model_checkpoint_path: "inference_model.ckpt"
```
Comment (jiny2001) [inline models/checkpoint:1]: It looks this file is committed by mistake. Actually we should remove this file from the repository.

File: evaluate.py [hunk 12]
```diff
@@ -65,7 +65,7 @@ def main(not_parsed_args):
         for test_data in test_list:
             evaluate_model(model, test_data)
 
-    model.save_model(name="inference_model", output_log=True, eval=False)
+    model.save_model(name="inference_model", output_log=True)
```
Comment (jiny2001) [inline evaluate.py:68]: Please don't use fixed filename, "inference_model".

File: models/checkpoint (hunk 11 above)
Comment (jiny2001) [reply models/checkpoint:1]: Also please exclude those private model files (.ckpt.*)
If we the model looks good and you want to include the model, I would like to check the performance. ;)

**Comment by tehtea** [comment]
Hi Jin, thanks for reviewing, I will edit the PR accordingly and I'll re-commit in a few day's time.

**Comment by jiny2001** [comment]
Hi, sorry for the many comments. 
I really appreciate your commit! However, I need to keep the repository clean.

So, how about creating more small commits? I guess you can create separate commits like one is for depthwise_seperable_conv2d, and the other is for creating frozen_graph or inference. The we can merge it step by step.

**Comment by tehtea** [comment]
Don't worry! Your reviews will help me to improve as well... please leave this PR open for now as I split it up accordingly.

**Pull request closed without merging**